src/runtime_src/core/common/drv/xrt_xclbin.c

```diff
@@ -0,0 +1,267 @@
+
+	/* Get byte length of bitstream */
+	header->BitstreamLength = data[index++];
+	header->BitstreamLength = (header->BitstreamLength << 8) | data[index++];
+	header->BitstreamLength = (header->BitstreamLength << 8) | data[index++];
+	header->BitstreamLength = (header->BitstreamLength << 8) | data[index++];
+	header->HeaderLength = index;
+
+	return 0;
+}
+
+void
+xrt_xclbin_free_header(struct XHwIcap_Bit_Header *header)
+{
+	vfree(header->DesignName);
+	vfree(header->PartName);
+	vfree(header->Date);
+	vfree(header->Time);
+}
+
+char *
+xrt_xclbin_kind_to_string(enum axlf_section_kind kind)
+{
+	switch (kind) {
+	case 0:  return "BITSTREAM";
+	case 1:  return "CLEARING_BITSTREAM";
+	case 2:  return "EMBEDDED_METADATA";
+	case 3:  return "FIRMWARE";
+	case 4:  return "DEBUG_DATA";
+	case 5:  return "SCHED_FIRMWARE";
+	case 6:  return "MEM_TOPOLOGY";
+	case 7:  return "CONNECTIVITY";
+	case 8:  return "IP_LAYOUT";
+	case 9:  return "DEBUG_IP_LAYOUT";
+	case 10: return "DESIGN_CHECK_POINT";
+	case 11: return "CLOCK_FREQ_TOPOLOGY";
+	case 12: return "MCS";
+	case 13: return "BMC";
+	case 14: return "BUILD_METADATA";
+	case 15: return "KEYVALUE_METADATA";
+	case 16: return "USER_METADATA";
+	case 17: return "DNA_CERTIFICATE";
+	case 18: return "PDI";
+	case 19: return "BITSTREAM_PARTIAL_PDI";
+	case 20: return "DTC";
+	case 21: return "EMULATION_DATA";
+	case 22: return "SYSTEM_METADATA";
+	case 23: return "SOFT_KERNEL";
+	case 24: return "ASK_FLASH";
+	case 25: return "AIE_METADATA";
+	case 26: return "ASK_GROUP_TOPOLOGY";
+	case 27: return "ASK_GROUP_CONNECTIVITY";
+	default: return "UNKNOWN";
+	}
+}
+
+const struct axlf_section_header *
+xrt_xclbin_get_section_hdr(const struct axlf *xclbin, enum axlf_section_kind kind)
+{
+	int i = 0;
+
+	for (i = 0; i < xclbin->m_header.m_numSections; i++) {
+		if (xclbin->m_sections[i].m_sectionKind == kind)
+			return &xclbin->m_sections[i];
+	}
+
+	return NULL;
+}
+
+int
+xrt_xclbin_check_section_hdr(const struct axlf_section_header *header,
+	uint64_t xclbin_len)
+{
+	return (header->m_sectionOffset + header->m_sectionSize) > xclbin_len ?
+		-EINVAL : 0;
+}
+
+int
+xrt_xclbin_section_info(const struct axlf *xclbin, enum axlf_section_kind kind,
+	uint64_t *offset, uint64_t *size)
+{
+	const struct axlf_section_header *memHeader = NULL;
+	uint64_t xclbin_len;
+	int err = 0;
+
+	memHeader = xrt_xclbin_get_section_hdr(xclbin, kind);
+	if (!memHeader)
+		return -EINVAL;
+
+	xclbin_len = xclbin->m_header.m_length;
+	err = xrt_xclbin_check_section_hdr(memHeader, xclbin_len);
+	if (err)
+		return err;
+
+	*offset = memHeader->m_sectionOffset;
+	*size = memHeader->m_sectionSize;
+
+	return 0;
+}
+
+int xrt_xclbin_get_section(const struct axlf *xclbin,
+	enum axlf_section_kind kind, void **data, uint64_t *len)
+{
+	void *section = NULL;
+	int err = 0;
+	uint64_t offset = 0;
+	uint64_t size = 0;
+
+	err = xrt_xclbin_section_info(xclbin, kind, &offset, &size);
+	if (err)
+		return err;
+
+	section = vmalloc(size);
```

Comment:
Seems like the caller should free this memory. Could we add a comment here?
Or maybe we could add a function "xrt_xclbin_put_section()" for release.